client/my-sites/people/role-select/index.jsx

```diff
@@ -2,138 +2,87 @@
-
-		if ( siteId ) {
-			let siteRoles = RolesStore.getRoles( siteId );
-
-			if ( this.props.includeFollower ) {
-				siteRoles = Object.assign( {}, this.getWPCOMFollowerRole(), siteRoles );
-			}
-
-			this.setState( {
-				roles: siteRoles
-			} );
-		}
-	},
-
-	fetchRoles( siteId = this.props.siteId ) {
-		if ( ! siteId ) {
-			debug( 'siteId not set' );
-			return;
-		}
-
-		if ( Object.keys( RolesStore.getRoles( siteId ) ).length ) {
-			debug( 'initial fetch not necessary' );
-			return;
-		}
-
-		debug( 'Fetching roles for ' + siteId );
+import QuerySites from 'components/data/query-sites';
+import QuerySiteRoles from 'components/data/query-site-roles';
+import { getSite } from 'state/sites/selectors';
+import { getSiteRoles } from 'state/site-roles/selectors';
+
+const getWPCOMFollowerRole = ( { site, translate } ) => {
+	const displayName = site.is_private
+		? translate( 'Viewer', { context: 'Role that is displayed in a select' } )
+		: translate( 'Follower', { context: 'Role that is displayed in a select' } );
+
+	return {
+		display_name: displayName,
+		name: 'follower'
+	};
+};
+
+const RoleSelect = ( props ) => {
+	const { siteRoles, site, includeFollower, siteId, id, explanation, translate } = props;
+	const roles = siteRoles && siteRoles.slice( 0 );
+	const omitProps = [
+		'site',
+		'key',
+		'siteId',
+		'includeFollower',
+		'explanation',
+		'siteRoles',
+		'dispatch',
+		'moment',
+		'numberFormat',
+		'translate'
+	];
+
+	if ( site && roles && includeFollower ) {
+		roles.push( getWPCOMFollowerRole( props ) );
+	}
 
-		// defer fetch requests to avoid dispatcher conflicts
-		setTimeout( function() {
-			const fetching = RolesStore.isFetching( siteId );
-			if ( fetching ) {
-				return;
+	return (
+		<FormFieldset key={ siteId } disabled={ ! roles }>
+			{ siteId &&
+				<QuerySites siteId={ siteId } /> &&
+				<QuerySiteRoles siteId={ siteId } />
 			}
-			RolesActions.fetch( siteId );
-		}, 0 );
-	},
-
-	render() {
-		const roleKeys = Object.keys( this.state.roles );
-		return (
-			<FormFieldset key={ this.props.siteId } disabled={ ! roleKeys.length }>
-				<FormLabel htmlFor={ this.props.id }>
-					{ this.translate( 'Role', {
-						context: 'Text that is displayed in a label of a form.'
-					} ) }
-				</FormLabel>
-				<FormSelect { ...omit( this.props, [ 'site', 'key', 'siteId', 'includeFollower', 'explanation' ] ) }>
-					{
-						map( this.state.roles, ( roleObject, key ) => {
-							return (
-								<option value={ key } key={ key }>
-									{ roleObject.display_name }
-								</option>
-							);
-						} )
-					}
-				</FormSelect>
-				{ this.props.explanation &&
-					<FormSettingExplanation>
-						{ this.props.explanation }
-					</FormSettingExplanation>
+			<FormLabel htmlFor={ id }>
+				{ translate( 'Role', {
+					context: 'Text that is displayed in a label of a form.'
+				} ) }
+			</FormLabel>
+			<FormSelect { ...omit( props, omitProps ) }>
+				{
+					roles && roles.map( ( role ) => {
```

Comment:
The nice thing about Lodash's `map` is that it would handle the empty `roles` for you automatically.

[It's also generally faster](https://jsperf.com/js-map-vs-lodash-map).

I don't feel particularly strongly about changing this.